src/lib/ValidityLib.sol

```diff
@@ -87,6 +90,35 @@ library ValidityLib {
         }
     }
 
+    /**
+     * @notice Internal function that validates a signature against an expected signer.
+     * should the initial verification fail, the SignatureDelegator is used to valdiate the claim
+     * Returns if the signature is valid or if the caller is the expected signer, otherwise
+     * reverts. The message hash is combined with the domain separator before verification.
+     * If ECDSA recovery fails, an EIP-1271 isValidSignature check is performed.
+     * If EIP-1271 fails, and an IEmissary is set for the sponsor, an IEmissary.verifyClaim check is performed
+     * @param messageHash     The EIP-712 hash of the message to verify.
+     * @param expectedSigner  The address that should have signed the message.
+     * @param signature       The signature to verify.
+     * @param domainSeparator The domain separator to combine with the message hash.
+     */
+    function signedBySponsorOrEmissary(bytes32 messageHash, address expectedSigner, bytes calldata signature, bytes32 domainSeparator, uint256[2][] memory idsAndAmounts) internal view {
+        // Apply domain separator to message hash and verify it was signed correctly.
+        bytes32 claimHash = messageHash.withDomain(domainSeparator);
+        // first check signature with ECDSA / ERC1271
+        // if the signature validation failed, fallback to emissary
+        bool hasValidSigner = expectedSigner.isValidSignatureNowCalldata(claimHash, signature) || claimHash.verifyWithEmissary(expectedSigner, idsAndAmounts.extractSameLockTag(), signature);
+
+        assembly ("memory-safe") {
+            // Allow signature check to be bypassed if caller is the expected signer.
```

Comment:
we might want to hoist the caller == signer check up so that the isValidSignatureNowCalldata check is skipped in those cases